bridges/otelzap/encoder.go

```diff
@@ -0,0 +1,266 @@
+	objectEncoderPool.Put(e)
+}
+
+var (
+	_ zapcore.ObjectEncoder = (*objectEncoder)(nil)
+	_ zapcore.ArrayEncoder  = (*arrayEncoder)(nil)
+)
+
+// Object Encoder implements zapcore.ObjectEncoder.
+// It encodes given fields to OTel attribute.
+type objectEncoder struct {
+	cur        []log.KeyValue
+	ctxfield   context.Context
+	reflectval log.Value
+}
+
+// NewObjectEncoder returns new ObjectEncoder.
+func newObjectEncoder(len int) *objectEncoder {
+	m := make([]log.KeyValue, 0, len)
+	return &objectEncoder{
+		cur: m,
+	}
+}
+
+// Converts Array to logSlice using ArrayEncoder.
+func (m *objectEncoder) AddArray(key string, v zapcore.ArrayMarshaler) error {
+	arr := getArrayEncoder()
+	err := v.MarshalLogArray(arr)
+	m.cur = append(m.cur, log.Slice(key, arr.elems...))
+	putArrayEncoder(arr)
+	return err
+}
+
+// Converts Object to logMap.
+func (m *objectEncoder) AddObject(k string, v zapcore.ObjectMarshaler) error {
+	newobj := getObjectEncoder()
+	err := v.MarshalLogObject(newobj)
+	m.cur = append(m.cur, log.Map(k, newobj.cur...))
+	putObjectEncoder(newobj)
+	return err
+}
+
+// Converts Binary to logBytes.
+func (m *objectEncoder) AddBinary(k string, v []byte) {
+	m.cur = append(m.cur, log.Bytes(k, v))
+}
+
+// Converts ByteString to logString.
+func (m *objectEncoder) AddByteString(k string, v []byte) {
+	m.cur = append(m.cur, log.String(k, string(v)))
+}
+
+// Converts Bool to logBool.
+func (m *objectEncoder) AddBool(k string, v bool) {
+	m.cur = append(m.cur, log.Bool(k, v))
+}
+
+// Converts Duration to logInt.
+func (m *objectEncoder) AddDuration(k string, v time.Duration) {
+	m.AddInt64(k, v.Nanoseconds())
+}
+
+// Converts Complex128 to logString.
+func (m *objectEncoder) AddComplex128(k string, v complex128) {
+	stringValue := strconv.FormatComplex(v, 'f', -1, 64)
+	m.cur = append(m.cur, log.String(k, stringValue))
+}
+
+// Converts Float64 to logFloat64.
+func (m *objectEncoder) AddFloat64(k string, v float64) {
+	m.cur = append(m.cur, log.Float64(k, v))
+}
+
+// Converts Float32 to logFloat64.
+func (m *objectEncoder) AddFloat32(k string, v float32) {
+	m.AddFloat64(k, float64(v))
+}
+
+// Converts Int64 to logInt64.
+func (m *objectEncoder) AddInt64(k string, v int64) {
+	m.cur = append(m.cur, log.Int64(k, v))
+}
+
+// Converts Int to logInt.
+func (m *objectEncoder) AddInt(k string, v int) {
+	m.cur = append(m.cur, log.Int(k, v))
+}
+
+// Converts String to logString.
+func (m *objectEncoder) AddString(k string, v string) {
+	m.cur = append(m.cur, log.String(k, v))
+}
+
+// Converts Uint64 to logInt64/logString
+func (m *objectEncoder) AddUint64(k string, v uint64) {
+	m.cur = append(m.cur,
+		log.KeyValue{
+			Key:   k,
+			Value: assignUintValue(v),
+		})
+}
+
+// Converts all non-primitive types to JSON string
+// Also checks for explcit context passed
+func (m *objectEncoder) AddReflected(k string, v interface{}) error {
+	if ctx, ok := v.(context.Context); ok {
+		// assign ctx
+		m.ctxfield = ctx
+		return nil
+	}
```

Comment:
Why not doing a type assertion like this in `getAttr`?

Reply:
See https://github.com/open-telemetry/opentelemetry-go-contrib/pull/5279#discussion_r1544294319